Parsing for Skill Multipliers (Both Leader and Active)

rohilthopu opened the pull request:
I know this already exists in Rikuu, but I was gonna use this for my own purposes, but it fit nicely into the setup you already have. 

To get the max multiplier for, say, a leader skill, you would use the multiplier listed on that skill times the multipliers of the skills in skill_part_1 to skill_part_3. 

This also includes multipliers for active skills, i.e. 2.5x for some type, has an atk_mult of 2.5, but the rest are either 1 or 0 (in the case of a shield). 

Shields are listed as fractional, i.e. 75% damage reduction is 0.75.

Could use a ton of refactoring.

I did not change any of the values currently available, but only added new ones. 

If you want to use this, you're more than welcome to. Regardless I'm using something similar for my own purposes.

nachoapps commented:
Thanks! This is great; we already have a really ugly skill parser that does LS multiplier calculations but the code is such a wreck i didn't want to integrate it more centrally.

The one thing that code does have going for it is that it's validated. Have you compared the results of this code to the results from that code? 

I would do it myself but I'm pretty focused on PadGuide for now since our ugly solution is technically working.

Assuming you have validated the results I'd be happy to merge this.

rohilthopu commented:
I've checked a ton of values against their actuals and as far as my not so perfect test suite goes, I think i've squashed most, if not all, of the bugs. To give you an idea of what the data looks like (when multiplied across skills), I've attached some images with the leader skill calculations from my own site that I'm working on. They have calculations done for a bunch of different types of leaders skills.

I'll keep checking and will push again if I find anything else, but from what I've gone through, I don't see any errors currently.

![screen shot 2018-08-05 at 12 14 26 am](https://user-images.githubusercontent.com/19541116/43682805-b5aa9f66-9844-11e8-83e2-87ae1c6bd080.png)
![screen shot 2018-08-05 at 12 14 11 am](https://user-images.githubusercontent.com/19541116/43682806-b5d595e0-9844-11e8-8dbb-09f07c0101c5.png)
![screen shot 2018-08-05 at 12 13 49 am](https://user-images.githubusercontent.com/19541116/43682807-b5e341b8-9844-11e8-99b8-bdc1bf9c1864.png)
![screen shot 2018-08-05 at 12 13 04 am](https://user-images.githubusercontent.com/19541116/43682808-b5f13322-9844-11e8-832a-b6519caac9b8.png)
![screen shot 2018-08-05 at 12 08 38 am](https://user-images.githubusercontent.com/19541116/43682809-b5ff00b0-9844-11e8-856d-6743441762c0.png)
![screen shot 2018-08-05 at 12 08 15 am](https://user-images.githubusercontent.com/19541116/43682810-b60e2928-9844-11e8-83b5-bfa9d6a9e8b0.png)
![screen shot 2018-08-05 at 12 07 58 am](https://user-images.githubusercontent.com/19541116/43682811-b626203c-9844-11e8-941d-5e17f099a537.png)
![screen shot 2018-08-05 at 12 07 24 am](https://user-images.githubusercontent.com/19541116/43682812-b6331256-9844-11e8-8876-4e7e443bf7b1.png)
![screen shot 2018-08-05 at 12 01 24 am](https://user-images.githubusercontent.com/19541116/43682813-b64083dc-9844-11e8-908c-6f3f055e0429.png)

nachoapps commented:
It's great to spot check, but there's an easier way. Just generate the values for each monster sequentially, dump them to text files, and diff the files. 

This data format isn't exactly super clear; it's pretty easy to make mistakes.

rohilthopu commented:
I meant to respond to this earlier but, I actually did run it against the actuals. My bad, I should've included those results too. I dumped the results to the txt attached here. The values that fail here are almost all because I'm apparently incapable of multiplying multiple shields (though their individual skill multipliers are correct) or they are values that are either, correct on my end but wrong on yours, or values that I would argue are more correct by my methods.

[results.txt](https://github.com/nachoapps/rpad-cogs-utils/files/2260872/results.txt)

nachoapps commented:
Shields are just: `(1 - (1 - resist1) * (1 - resist2))`

Thanks for catching Tifa, cate is fixing.

It looks like besides the shields the only other confirmed problem is Wallace, looks pretty good.

I think I disagree with your approach to apply the optional negative modifiers. People are going to build teams around not hitting these modifiers.

rohilthopu commented on pad_api_data/pad_etl/common/pad_util.py:
```diff
@@ -81,3 +81,489 @@ def get_dungeon_comment(val: int) -> str:
+
+def parse_skill_multiplier(skill, other_fields, length) -> {}:
+    # HP, ATK, RCV, Damage Reduction
+    multipliers = {'hp': 1.0, 'atk': 1.0, 'rcv': 1.0, 'shield': 0.0}
```
Yeah that would be a good idea. Definitely improves the readability of this mess of if/else's i created. I'll get on that when I get out of work later. Sorry for the delay.

rohilthopu commented:
I went ahead and changed the things you recommended. I also switched the multipliers to always take the best case. The only area where I disagree is the calculation of skill 8896 from my results.txt. I'm not sure how or why, but using your parser gets a 0x multiplier for RCV when there is literally no data to draw that conclusion. It is also the only skill for its skill type.

dchang0413 commented:
I need to point out, ID 8092, leader skill of Lucifer has 2 color crosses and heart cross, the actual max ATK multiplier should be 625x rather than 3903.75x, it's an existing problem on my code that I haven't visited

nachoapps commented:
Looks good, thanks! I'll probably swap over to your multiplier calculation and purge the other one since it's such a mess.

rohilthopu commented:
@dchang0413 Yeah that's my bad as well. I didn't catch that only 3 cross (one heard, two other attr) is possible. I'll correct it on my end as well asap.
